test/optimus/scientific/ScientificTests.wdl

```diff
@@ -1,26 +1,70 @@
+task TestBam {
+
+  # Compare the tagged/aligned bam generated by Optimus to expected values derived from running Cell Ranger on the same data"
 
-# note that this is not a great test because it localizes the .bam file
-# however it is a good placeholder while we put together the scientific tests
-# themselves, as these files will be relatively small in the testing data.
-task TestBamRecordNumber {
   File bam
-  Int expected_records
+
+  Int expected_records = 3055752
+  Int chr1_expected = 1170508
+  Int chr2_expected = 134053
+  Int unmapped_expected = 134321
+
   Int required_disk = ceil(size(bam, "G") * 1.2)
 
+  parameter_meta {
+    bam: "The aligned and tagged bamfile from the Optimus workflow, run on synthetic data"
+    expected_records: "Total number of records in bam"
+    chr1_expected: "Total number of expected alignments for chromosome 1"
+    chr2_expected: "Total number of expected alignments for chromosome 2"
+    unmapped_expected: "Total number of unmapped alignments"
+  }
+
   command {
 
     # catch uncommon errors
-    -eo pipefail
+    set -eo pipefail
+
+    # we don't have an index, so generate one here
+    samtools index "${bam}"
 
     N_RECORDS=$(samtools view "${bam}" | wc -l)
     if [ $N_RECORDS != "${expected_records}" ]; then
         >&2 echo "Number of records in the pipeline output ($N_RECORDS) did not match the expected number (${expected_records})"
         exit 1
     fi
+
+    python3 <<CODE
+
+    import pysam
+    import sys
+
+    bamfile = '${bam}'
+    bam = pysam.AlignmentFile(bamfile)
+
+    chr1_records = sum(1 for _ in bam.fetch(contig='1'))
+    chr2_records = sum(1 for _ in bam.fetch(contig='2'))
+    unmapped_records = sum(1 for record in bam if record.is_unmapped)
+
+    chr1_expected = ${chr1_expected}
+    if chr1_records != chr1_expected:
+        sys.stderr.write('chr1 observed records %d != expected records %d' % (chr1_records, chr1_expected))
+        sys.exit(1)
+
+    chr2_expected = ${chr2_expected}
+    if chr2_records != chr2_expected:
+        sys.stderr.write('chr2 observed records %d != expected records %d' % (chr2_records, chr2_expected))
+        sys.exit(1)
+
+    unmapped_expected = ${unmapped_expected}
+    if unmapped_records != unmapped_expected:
+        sys.stderr.write('unmapped observed records %d != expected records %d' % (unmapped_records, unmapped_expected))
+        sys.exit(1)
+
```

Comment:
What do you think about exiting at the end if there are any errors, instead of after each one? If there are multiple errors you could then see all of them instead of failing after the first error.

Reply:
Yep, good idea.